spirv_msl.cpp

```diff
@@ -1588,6 +1589,25 @@ void CompilerMSL::preprocess_op_codes()
 	                          (need_subpass_input_ms && !msl_options.use_framebuffer_fetch_subpasses))))
 		needs_sample_id = true;
 
+	// OpKill is removed by the parser, so we need to identify those by inspecting
+	// blocks.
+	ir.for_each_typed_id<SPIRBlock>([&preproc](uint32_t, SPIRBlock &block) {
+		if (block.terminator == SPIRBlock::Kill)
+			// FIXME: This doesn't distinguish between OpKill and OpTerminateInvocation!
+			preproc.uses_discard = true;
+	});
+
+	// Fragment shaders that both write to storage resources and discard fragments
+	// need checks on the writes, to work around Metal allowing these writes despite
+	// the fragment being dead. Similarly, fragment shaders that use OpIsHelperInvocationEXT,
+	// or read a volatile BuiltInHelperInvocation, need this extra check, because
+	// despite that keeping derivatives well-defined is the sole purpose of allowing a
+	// fragment thread to run after discard, simd_is_helper_thread() doesn't return
+	// true for discarded fragment threads.
```

Comment:
... what?

Reply:
Yes, you read that right.

Reply:
Well, that part of the comment's gone now, since I split this up into two options, but my point still stands.